File: LeetCode/ProblemSet/012.Integer to Roman/integer-to-roman.js
/**
 * @param {number} num
 * @return {string}
 */
var intToRoman = function (num) {
    let map = {
        1: 'I',
        4: 'IV',
        5: 'V',
        9: 'IX',
        10: 'X',
        50: 'L',
        40: 'XL',
        90: 'XC',
        100: 'C',
        400: 'CD',
        500: 'D',
        900: 'CM',
        1000: 'M'
    };

    let keys = Object.keys(map);

    let r = '';
    while (num > 0) {
        let start = num >= 50 ? keys.length - 1 : 6;
        for (let i = start; i >= 0; i--) {
            if (num >= keys[i]) {
                num -= keys[i];
                r += map[keys[i]];
                break;
            }
        }
    }
    return r;
};

console.log(intToRoman(3));
console.log(intToRoman(4));
console.log(intToRoman(49));
console.log(intToRoman(58));
console.log(intToRoman(1994));

File: LeetCode/Go-Solutions/ProblemSet/0012.Integer_to_Roman/integer-to-roman.go
package _012_Integer_to_Roman

func intToRoman(num int) string {
	m := []struct {
		n int
		s string
	}{
		{1000, "M"},
		{900, "CM"},
		{500, "D"},
		{400, "CD"},
		{100, "C"},
		{90, "XC"},
		{50, "L"},
		{40, "XL"},
		{10, "X"},
		{9, "IX"},
		{5, "V"},
		{4, "IV"},
		{1, "I"},
	}

	l, i := len(m), 0
	r := ""
	for num > 0 {
		i = 0
		if num < 90 {
			i = 6
		}
		for ; i < l; i++ {
			if num >= m[i].n {
				num -= m[i].n
				r += m[i].s
				break;
			}
		}
	}
	return r
}
